src/license_guc.c

```diff
@@ -71,315 +43,218 @@ current_license_can_downgrade_to_apache(void)
+license_type_of(const char *string)
+{
+	if (string == NULL)
+		return LICENSE_UNDEF;
+	if (strcasecmp(string, TS_LICENSE_TSL) == 0)
+		return LICENSE_TSL;
+	if (strcasecmp(string, TS_LICENSE_APACHE) == 0)
+		return LICENSE_APACHE;
+	return LICENSE_UNDEF;
 }
 
-/*
- * `ts_license_update_check`
- * Used to validate license keys in preparation for `ts_license_on_assign`
- */
-bool
-ts_license_update_check(char **newval, void **extra, GucSource source)
+TSDLLEXPORT bool
+ts_license_is_apache(void)
 {
-	Datum module_can_start;
-	bool try_to_load_tsl = true;
-
-	if (*newval == NULL)
-		return false;
-
-	if (!TS_LICENSE_TYPE_IS_VALID(*newval))
-		return false;
-
-	/*
-	 * we can shutdown the submodule if it was never loaded, or it has a valid
-	 * shutdown function.
-	 */
-	if (TS_LICENSE_IS_APACHE_ONLY(*newval))
-	{
-		if (!current_license_can_downgrade_to_apache())
-		{
-			GUC_check_errdetail("Cannot downgrade a running session to Apache Only.");
-			GUC_check_errhint("change the license in the configuration file");
-			return false;
-		}
-
-		try_to_load_tsl = false;
-	}
-
-	if (!can_load)
-	{
-		load_source = source;
-		return true;
-	}
+	return license_type_of(ts_guc_license) == LICENSE_APACHE;
+}
 
-	if (!try_to_load_tsl)
+static bool
+license_is_downgradable(void)
+{
+	if (downgrade_enabled)
 		return true;
 
-	if (!load_tsl())
-	{
-		GUC_check_errdetail("Could not find additional timescaledb module");
-		GUC_check_errhint("check that %s-%s is available",
-						  TSL_LIBRARY_NAME,
-						  TIMESCALEDB_VERSION_MOD);
-		return false;
-	}
-
-	Assert(tsl_handle != NULL);
-	Assert(tsl_validate_license_fn != NULL);
-	Assert(tsl_startup_fn != NULL);
-#ifdef WIN32
-
-	/*
-	 * freeing the guc extra causes heap corruption on windows so instead we
-	 * re-parse in the assign hook
-	 */
-	extra = NULL;
-#endif
-
-	module_can_start = DirectFunctionCall2(tsl_validate_license_fn,
-										   CStringGetDatum(*newval),
-										   PointerGetDatum(extra));
-
-	return DatumGetBool(module_can_start);
+	return ts_license_is_apache() && tsl_handle == NULL;
 }
 
-#ifdef WIN32
-static void *
-revalidate_license(const char *newval)
+TSDLLEXPORT void
+ts_license_enable_downgrade(void)
 {
-	void *extra = NULL;
-	void **extra_p = &extra;
-
-	Assert(extra == NULL);
-
 	/*
-	 * Due to windows issues we cannot use the extra parameter. Instead
-	 * re-call the validation function, if we reach this point the license
-	 * must be valid so the function cannot fail cannot fail.
+	 * For testing purposes we occasionally need the ability to set the license to
+	 * Apache only. This function allows us to bypass the test that usually disables
+	 * that.
 	 */
-	DirectFunctionCall2(tsl_validate_license_fn, CStringGetDatum(newval), PointerGetDatum(extra_p));
-	return extra;
+	downgrade_enabled = true;
 }
-#endif
 
-/*
- * `ts_license_on_assign`
- * Links the cross-module function pointers.
- */
-void
-ts_license_on_assign(const char *newval, void *extra)
+TSDLLEXPORT void
```

Comment:
Why does this need to be exported? I see it called from TSL, which seems really strange given that loading must already be enabled when called from TSL, otherwise the library wouldn't be loaded. 

I realize this was there before, but the whole thing seems quite intuitive.